### PR TITLE
Not trigger channelWritabilityChanged if fail messages before close Chan...

### PR DESCRIPTION
**Pull request by normanmaurer**
...nel.

Motivation:

We should not trigger channelWritabilityChanged during failing message when we are about to close the Channel as otherwise the use may try again writing even if the Channel is about to get closed.

Modifications:

Add new boolean param to ChannelOutboundBuffer.failFlushed(...) which allows to specify if we should notify or not.

Result:

channelWritabilityChanged is not triggered anymore if we cloe the Channel because of an IOException during write.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin please review

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 1]
```diff
@@ -573,7 +577,7 @@ public boolean isEmpty() {
         return flushed == 0;
     }
 
-    void failFlushed(Throwable cause) {
+    void failFlushed(Throwable cause, boolean notify) {
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: `notify` -> `notifyWritability` for clarity

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 2]
```diff
@@ -759,8 +760,10 @@ protected void flush0() {
             try {
                 doWrite(outboundBuffer);
             } catch (Throwable t) {
-                outboundBuffer.failFlushed(t);
-                if (t instanceof IOException && config().isAutoClose()) {
+                boolean close = t instanceof IOException && config().isAutoClose();
+                // We not want to notify for pendingWritableBytes updates if we close the channel anyway.
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: 'We do not want to trigger channelWritabilityChanged event if the channel is going to be closed.`

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 3]
```diff
@@ -746,9 +746,10 @@ protected void flush0() {
             if (!isActive()) {
                 try {
                     if (isOpen()) {
-                        outboundBuffer.failFlushed(NOT_YET_CONNECTED_EXCEPTION);
+                        outboundBuffer.failFlushed(NOT_YET_CONNECTED_EXCEPTION, true);
                     } else {
-                        outboundBuffer.failFlushed(CLOSED_CHANNEL_EXCEPTION);
+                        // Not notify as the channel is closed already.
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: 'Do not trigger channelWritabilityChanged because the channel is closed already.'

**Comment by trustin** [comment]
Please feel free to cherry-pick after addressing the comments.

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java (hunk 1 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java]: done

File: transport/src/main/java/io/netty/channel/AbstractChannel.java (hunk 2 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: done

File: transport/src/main/java/io/netty/channel/AbstractChannel.java (hunk 3 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: done

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (0d792c3aa0d49434cbed90a6d136e1440df56a31) , 4.1 (57e7571c65f181df97ea7bbab372659a0baedfa1)  and master (0615e538c17ecc0191acc65978f2d6b01b160a19)

**Pull request closed without merging**